File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/merge_k_sorted_lists.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;

  ListNode(int val) : val(val), next(nullptr) {}
};

void display(ListNode *head) {
  while (head != nullptr) {
    cout << head->val << " ";
    head = head->next;
  }
  cout << "\n";
}

struct CompareNode {
  bool operator()(const ListNode *n1, ListNode *n2) {
    return n1->val > n2->val;
  }
};

// O(nlogk) / O(k)
ListNode *merge_k_sorted_lists(const vector<ListNode *> &lists) {
  priority_queue<ListNode *, vector<ListNode *>, CompareNode> min_heap;

  for (ListNode *head : lists) {
    if (head != nullptr) {
      min_heap.push(head);
    }
  }

  ListNode *res_head = nullptr, *tail = nullptr;
  while (!min_heap.empty()) {
    ListNode *node = min_heap.top();
    min_heap.pop();

    if (res_head == nullptr) {
      res_head = tail = new ListNode(node->val);
    } else {
      tail->next = new ListNode(node->val);
      tail = tail->next;
    }

    if (node->next != nullptr) {
      min_heap.push(node->next);
    }
  }

  return res_head;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m, val;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<ListNode *> lists;
    for (int i = 0; i < n; i++) {
      cin >> m;

      ListNode *head = nullptr, *tail = nullptr;
      for (int i = 0; i < m; i++) {
        cin >> val;
        if (head == nullptr) {
          head = tail = new ListNode(val);
        } else {
          tail->next = new ListNode(val);
          tail = tail->next;
        }
      }
      lists.emplace_back(head);
    }

    auto res_head = merge_k_sorted_lists(lists);
    display(res_head);
  }

  return 0;
}

File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/MergeKSortedLists.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MergeKSortedLists {
  static class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
      this.val = val;
      this.next = null;
    }
  }

  private static void display(ListNode head) {
    while (head != null) {
      System.out.printf("%d ", head.val);
      head = head.next;
    }
    System.out.println();
  }

  // O(nlogk) / O(k)
  public static ListNode mergeKSortedLists(List<ListNode> lists) {
    PriorityQueue<ListNode> minHeap = new PriorityQueue<>(
        (ListNode l1, ListNode l2) -> Integer.compare(l1.val, l2.val));

    int k = lists.size();
    ListNode resHead = null, resTail = null;

    for (int i = 0; i < k; i++) {
      if (lists.get(i) != null) {
        minHeap.add(lists.get(i));
      }
    }

    while (!minHeap.isEmpty()) {
      ListNode node = minHeap.poll();

      if (resHead == null) {
        resHead = resTail = new ListNode(node.val);
      } else {
        resTail.next = new ListNode(node.val);
        resTail = resTail.next;
      }

      if (node.next != null) {
        node = node.next;
        minHeap.add(node);
      }
    }

    return resHead;
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<ListNode> lists = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          int m = in.nextInt();
          ListNode head = null, currHead = null;

          for (int j = 0; j < m; j++) {
            int val = in.nextInt();

            if (head == null) {
              head = currHead = new ListNode(val);
            } else {
              currHead.next = new ListNode(val);
              currHead = currHead.next;
            }
          }

          lists.add(head);
        }

        ListNode merged = mergeKSortedLists(lists);
        display(merged);
      }
    }
  }
}